### ISSUE TITLE
USWDS - Bug: Search input uses non-standard clear button that is not available in Firefox

### ISSUE THREAD

**Issue by anselmbradford** [uswds/uswds#5834]
### Describe the bug

After entering text in the search inputs on https://designsystem.digital.gov/components/search/, a clear button is shown. I have a few questions around this:

- This uses a non-standard browser default clear button that is not supported in Firefox, so Firefox visitors will get no clear button. 

- It also does not have a hover state.

- It is not accessible via the keyboard. I'm not sure if this is an issue or not. Is the expectation with a screenreader that the escape key could be used to clear the input, and it is therefore not necessary that the clear button be accessible?

### Steps to reproduce the bug

1. Go to https://designsystem.digital.gov/components/search/ in Chrome.
2. Enter some text in the search input and see there is a clear button that appears.
3. Go to the same page in Firefox.
4. Enter some text in the search input and see there is no clear button.

### Expected Behavior

Clear button is consistent across major browsers.

### Screenshots

Chrome
<img width="599" alt="Screenshot 2024-03-22 at 5 46 56 PM" src="https://github.com/uswds/uswds/assets/704760/cf0fced0-0e97-4861-b02f-ec767f7ba815">

Firefox
<img width="601" alt="Screenshot 2024-03-22 at 5 46 52 PM" src="https://github.com/uswds/uswds/assets/704760/8c8af23e-5623-422e-ad00-44e5fd48583a">

### Additional context

This branch shows how a cross-browser (x) clear button could be created via the use of a reset button accompanying the search input. However, this may not be an optimal solution https://github.com/cfpb/consumerfinance.gov/compare/ans_global_search_fix?expand=1

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Bug%22) for duplicate bug reports.

**Comment by mejiaj** [comment]
_Based on 04-09-24 Dev sync discussion..._

- [ ] A11Y team will test to determine priority & severity.
- [ ] Outline on site guidance that this is a browser issue.

**Comment by amycole501** [comment]
Using both JAWS and NVDA I don't hear the "clear" button announced. It's inoperable to both screen readers and keyboard users.

Firefox is suppressing the feature and this has been brought to the attention to other design systems. Also this https://github.com/jquery-archive/jquery-mobile/issues/6467

[User agents have to comply with WCAG 2.0]( https://www.w3.org/TR/UAAG20/#gl-keyboard-access).  I suggest the issue [go to Firefox](https://github.com/mozilla/).


For now, I don't recommend we prioritize a solution due to the fact that this is a single browser issue and there is an alternate means by which the user can delete content out of the field.

That said, I plan to audit the search component this sprint; I can further update this issue as my research develops.

**Comment by anselmbradford** [comment]
@amycole501 I took the reset button technique referenced above and slimmed it down to a jsfiddle at https://jsfiddle.net/9buarncv/7/ (viewable full screen at https://jsfiddle.net/9buarncv/7/show). How does that look for accessibility? My testing shows it working the same in Firefox and Chrome, but Safari does not allow tabbing to the button, which would be equivalent to the status quo of the default (x) button that shows up in Chrome. So cross-browser improvement in Chrome and Firefox, status quo in Safari (although it does make the appearance of the (x) in Safari consistent with the others).

**Comment by anselmbradford** [comment]
I built a more robust implementation based on the jsfiddle above (with some changes, such as using flexbox), which is being added to the CFPB design system in https://github.com/cfpb/design-system/pull/1979

Feedback welcome. It does include a small snippet of inline JS, because in cases where the value attribute is set by the server, the reset button will only return it to the state it was in on page load. Therefore, the JS completely clears the input for that circumstance. I figure if this will become a web component in the future, how it handles js/no-js will be buried in the component internals, so a small amount of inline JS is ok for now.

**Comment by mejiaj** [comment]
Removed assignees because we've confirmed the issue and have potential solutions, thanks @anselmbradford.

## Todo

- [ ] Engineering to look at proposed solutions
- [ ] Get the weight and complexity
- [ ] Discuss with team

**Comment by aduth** [comment]
Related: #5277

**Comment by mgifford** [comment]
I can confirm that the x still doesn't work in Firefox here, although it does in Chrome:
https://designsystem.digital.gov/components/search/

I noticed a problem that the "x" to clear the image isn't keyboard accessible.

However looking at other design sytems with this:
https://design.sis.gov.uk/components/inputs/search-bar/accessibility/

or even closer to home:
https://design.va.gov/components/search-input

The code proposed here also seems to work in Firefox and Chrome:
https://jsfiddle.net/9buarncv/7/

but It was too bad that the Search button wasn't also included in his example.

**Comment by chandracarney** [comment]
In June, we discussed that this needs further testing and triage. Here are our initial thoughts:

- When the issue is with the browser itself, an accessibility principle is generally that we shouldn't be touching built-in browser issues; instead open issues with the browser teams to pressure them to fix issues. (Browsers or any upstream dependency: hardware and software alike.)
- Need to read through all the links and background before a decision to close or proceed.
- If we make changes:
  - Danger: If the browsers update, the cascade to USWDS could conflict with our work and cause problems for sites running USWDS. Need to be careful.
  - We'd need to do some hacking to make it work as described.

**Comment by ryparker** [comment]
I'm closing this as not planned and continuing the proposal for an optional clear button in [discussion #6971](https://github.com/uswds/uswds/discussions/6971).

The CFPB implementation and the earlier JAWS/NVDA results are useful starting points. We still need to weigh the added markup and events, server-populated values, behavior without JavaScript, and the risk of conflicts with future browser changes. Standard keyboard text editing remains available; we haven't approved replacing the native control by default.

The discussion includes the proposal and compatibility requirements. GitHub's issue conversion didn't complete, so it links back here to preserve the original report and comments. The usability concern remains open for discussion.
